HoRNDIS.cpp

```diff
@@ -891,74 +858,80 @@ IOReturn HoRNDIS::message(UInt32 type, IOService *provider, void *argument) {
 }
+*/
 
 
 /***** Packet transmit logic *****/
 
+static inline bool isTransferStopStatus(IOReturn rc) {
+	// IOReturn indicating that we need to stop transfers:
+	return rc == kIOReturnAborted || rc == kIOReturnNotResponding;
+}
+
 UInt32 HoRNDIS::outputPacket(mbuf_t packet, void *param) {
-	mbuf_t m;
-	size_t pktlen = 0;
 	IOReturn ior = kIOReturnSuccess;
-	UInt32 poolIndx;
-	int i;
-
-	LOG(V_DEBUG, "");
+	int poolIndx = N_OUT_BUFS;
+
+	// Note, this function MAY or MAY NOT be protected by the IOCommandGate,
+	// depending on the kind of OutputQueue used.
+	// Here, we assume that IOCommandGate is used: no need to lock.
+
+	if (!fReadyToTransfer) {
+		// Technically, we must never be here, because we always disable the
+		// queue before clearing 'fReadyToTransfer', but double-checking here
+		// just in-case: better safe than sorry.
+		LOG(V_DEBUG, "fReadyToTransfer=false: dropping packet "
+			"(we shouldn't even be here)");
+		freePacket(packet);
+		return kIOReturnOutputDropped;
+	}
 	
-	/* Count the total size of this packet */
-	m = packet;
-	while (m) {
+	// Count the total size of this packet
+	size_t pktlen = 0;
+	for (mbuf_t m = packet; m; m = mbuf_next(m)) {
 		pktlen += mbuf_len(m);
-		m = mbuf_next(m);
 	}
 	
-	LOG(V_DEBUG, "%ld bytes", pktlen);
+	LOG(V_PACKET, "%ld bytes", pktlen);
+
+	const uint32_t transmitLength = (uint32_t)(pktlen + sizeof(rndis_data_hdr));
 	
-	if (pktlen > (mtu + 14)) {
-		LOG(V_ERROR, "packet too large (%ld bytes, but I told you you could have %d!)", pktlen, mtu);
+	if (transmitLength > maxOutTransferSize) {
+		LOG(V_ERROR, "packet too large (%ld bytes, maximum can transmit %ld)",
+			pktlen, maxOutTransferSize - sizeof(rndis_data_hdr));
 		fpNetStats->outputErrors++;
-		return false;
+		freePacket(packet);
+		return kIOReturnOutputDropped;
 	}
-	
-	/* Find an output buffer in the pool */
-	IOLockLock(outbuf_lock);
-	for (i = 0; i < OUT_BUF_MAX_TRIES; i++) {
-		uint64_t ivl, deadl;
-		
-		for (poolIndx = 0; poolIndx < N_OUT_BUFS; poolIndx++)
-			if (!outbufs[poolIndx].inuse) {
-				outbufs[poolIndx].inuse = true;
-				break;
-			}
-		if (poolIndx != N_OUT_BUFS)
-			break;
-		
-		/* "while", not "if".  See Symphony X's seminal work on this topic, /Paradise Lost/ (2007). */
-		nanoseconds_to_absolutetime(OUT_BUF_WAIT_TIME, &ivl);
-		clock_absolutetime_interval_to_deadline(ivl, &deadl);
-		LOG(V_NOTE, "waiting for buffer...");
-		
-		IOLockSleepDeadline(outbuf_lock, outbufs, *(AbsoluteTime *)&deadl, THREAD_INTERRUPTIBLE);
+
+	if (numFreeOutBufs <= 0) {
+		LOG(V_ERROR, "BUG: Ran out of buffers - stall did not work!");
+		// Stall the queue and re-try the same packet later: don't release:
+		return kIOOutputStatusRetry | kIOOutputCommandStall;
 	}
-	IOLockUnlock(outbuf_lock);
-	
-	if (poolIndx == N_OUT_BUFS) {
-		LOG(V_ERROR, "timed out waiting for buffer");
-		return kIOReturnTimeout;
+
+	// Note, we don't decrement 'numFreeOutBufs' (commit to using that buffer)
```

Comment:
And this is safe because we use a gated output queue.

Reply:
Note, I put the comment regarding gated output queue at the start of the function.